x/lockdrop/keeper/distr.go

```diff
@@ -0,0 +1,195 @@
+package keeper
+
+import (
+	"fmt"
+	"sort"
+
+	"github.com/osmosis-labs/osmosis/v15/osmoutils"
+
+	"github.com/dymensionxyz/dymension/x/lockdrop/types"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+)
+
+// FundCommunityPoolFromModule allows the lockdrop module to directly fund the community fund pool.
+func (k Keeper) FundCommunityPoolFromModule(ctx sdk.Context, asset sdk.Coin) error {
+	moduleAddr := k.accountKeeper.GetModuleAddress(types.ModuleName)
+	if moduleAddr == nil {
+		panic("Could not get distribution module from SDK")
+	}
+
+	return k.distrKeeper.FundCommunityPool(ctx, sdk.Coins{asset}, moduleAddr)
+}
+
+// AllocateAsset allocates and distributes coin according a gauge’s proportional weight that is recorded in the record.
+func (k Keeper) AllocateAsset(ctx sdk.Context) error {
+	logger := k.Logger(ctx)
+	params := k.GetParams(ctx)
+	moduleAddr := k.accountKeeper.GetModuleAddress(types.ModuleName)
+
+	//TODO(#5): we should support distribution of all assets, not only specific denom
+	asset := k.bankKeeper.GetBalance(ctx, moduleAddr, params.MintedDenom)
+	if asset.Amount.IsZero() {
+		// when allocating asset is zero, skip execution
+		return nil
+	}
+
+	distrInfo := k.GetDistrInfo(ctx)
+
+	if distrInfo.TotalWeight.IsZero() {
+		// If there are no records, put the asset to the community pool
+		return k.FundCommunityPoolFromModule(ctx, asset)
+	}
+
+	assetAmountDec := sdk.NewDecFromInt(asset.Amount)
+	totalWeightDec := sdk.NewDecFromInt(distrInfo.TotalWeight)
+	for _, record := range distrInfo.Records {
+		allocatingAmount := assetAmountDec.Mul(sdk.NewDecFromInt(record.Weight).Quo(totalWeightDec)).TruncateInt()
+
+		// when weight is too small and no amount is allocated, just skip this to avoid zero coin send issues
+		if !allocatingAmount.IsPositive() {
+			logger.Info(fmt.Sprintf("allocating amount for (%d, %s) record is not positive", record.GaugeId, record.Weight.String()))
+			continue
+		}
+
+		if record.GaugeId == 0 { // fund community pool if gaugeId is zero
+			if err := k.FundCommunityPoolFromModule(ctx, sdk.NewCoin(asset.Denom, allocatingAmount)); err != nil {
+				return err
+			}
+			continue
+		}
+
+		coins := sdk.NewCoins(sdk.NewCoin(asset.Denom, allocatingAmount))
+		err := k.incentivesKeeper.AddToGaugeRewards(ctx, k.accountKeeper.GetModuleAddress(types.ModuleName), coins, record.GaugeId)
```

Comment:
If one gauge id supplied is incorrect (i.e doesn't exist), the func will stop though it may fund the previous ones (assuming they were valid) . Seems like weird behavior to me. Shouldn’t we validate all gauges are valid/exist before starting to fund it?

Reply:
it's hard to check beforehand, as it may fail due to different reasons.
if it fails i'll move the funds to communityPool

Reply:
double checked it, 
it's already validated on `checkTx` in:

https://github.com/dymensionxyz/dymension/blob/0410b09fefb2a44370c3c0b0b204a129eee9f088/x/lockdrop/keeper/distr.go#L155